### PR TITLE
raftstore-v2: offload checkpoint during split

### LINKED ISSUE #14711: offload checkpoint during split

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #14711

What's Changed:

```commit-message
offload checkpoint during split
```
Compared to other apply tasks, rocksdb checkpoint is an expensive work in terms of CPU and IO where CPU works can block other tasks. 
Although pre-flush mechanism makes it much more lightweight, it is only a best effort and can still cause performance drops.
This PR offloads it to a dedicated worker so that the apply thread can execute other regions' tasks during the creation of the checkpoint.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by tonyxuqqi — APPROVED** [review]

**Comment by tonyxuqqi** [comment]
Please add the issue# and the explanation how does the PR work

**Comment by SpadeA-Tang** [comment]
> Please add the issue# and the explanation how does the PR work

Done

File: components/raftstore-v2/src/raft/apply.rs [hunk 1]
```diff
@@ -71,6 +72,8 @@ pub struct Apply<EK: KvEngine, R> {
     observe: Observe,
     coprocessor_host: CoprocessorHost<EK>,
 
+    checkpoint_cheduler: Scheduler<checkpoint::Task>,
```
Comment (bufferflies) [inline components/raftstore-v2/src/raft/apply.rs:75]:
```suggestion
    checkpoint_scheduler: Scheduler<checkpoint::Task>,
```

File: components/raftstore-v2/src/worker/checkpoint.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
+use std::{
```
Comment (overvenus) [inline components/raftstore-v2/src/worker/checkpoint.rs:1]: License header.

File: components/raftstore-v2/src/worker/checkpoint.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        // If it's recovered from restart, it's possible the target path exists already.
+        // And because checkpoint is atomic, so we don't need to worry about corruption.
+        // And it's also wrong to delete it and remake as it may has applied and flushed
+        // some data to the new checkpoint before being restarted.
+        if !derived_path.exists() {
+            checkpointer
+                .create_at(&derived_path, None, 0)
+                .unwrap_or_else(|e| {
+                    slog_panic!(
+                        self.logger,
```
Comment (overvenus) [inline components/raftstore-v2/src/worker/checkpoint.rs:111]: Does the logger have region id and peer id?

File: components/raftstore-v2/src/worker/checkpoint.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/checkpoint.rs:1]: Good catch.

File: components/raftstore-v2/src/worker/checkpoint.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/checkpoint.rs:115]: Good catch, I have added region id here.

File: components/raftstore-v2/src/worker/checkpoint.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                    )
+                });
+        }
+
+        let elapsed = now.saturating_elapsed();
+        // to be removed after when it's stable
+        info!(
+            self.logger,
+            "create checkpoint time consumes";
+            "region" =>  ?parent_region,
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/checkpoint.rs:129]:
```suggestion
            "region" => ?parent_region,
```

File: components/raftstore-v2/src/operation/command/admin/split.rs [hunk 5]
```diff
@@ -560,6 +532,25 @@ impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
...
 }
 
+async fn async_checkpoint(
+    scheduler: &Scheduler<checkpoint::Task>,
+    parent_region: u64,
+    split_regions: Vec<u64>,
+    cur_suffix: u64,
+    log_index: u64,
+) {
+    let (tx, rx) = oneshot::channel();
```
Comment (bufferflies) [inline components/raftstore-v2/src/operation/command/admin/split.rs:542]: how about return the executing duration and log it in the code line 500?

**Review by bufferflies — COMMENTED** [review]
rest LGTM

File: components/raftstore-v2/src/operation/command/admin/split.rs (hunk 5 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/split.rs:548]: Done

**Review by bufferflies — APPROVED** [review]

**Pull request merged**